src/sign/signatures/rrsigs.rs

```diff
@@ -0,0 +1,1840 @@
+                .rrsig_validity_period_strategy
+                .validity_period_for_rrset(&rrset);
+            let rrsig_rr = sign_rrset_in(
+                key.signing_key(),
+                &rrset,
+                zone_apex,
+                inception,
+                expiration,
+                reusable_scratch,
+            )?;
+            out.rrsigs.push(rrsig_rr);
+            trace!(
+                "Signed {} RRs in RRSET {} at the zone apex with keytag {}",
+                rrset.iter().len(),
+                rrset.rtype(),
+                key.signing_key().public_key().key_tag()
+            );
+        }
+    }
+
+    // Move the iterator past the processed apex owner RRs.
+    let _ = records.next();
+
+    Ok(())
+}
+
+/// Generate `RRSIG` records for a given RRset.
+///
+/// See [`sign_rrset_in()`].
+///
+/// If signing multiple RRsets, calling [`sign_rrset_in()`] directly will be
+/// more efficient as you can allocate the scratch buffer once and re-use it
+/// across multiple calls.
+pub fn sign_rrset<N, D, Octs, Inner>(
+    key: &SigningKey<Octs, Inner>,
+    rrset: &Rrset<'_, N, D>,
+    apex_owner: &N,
+    inception: Timestamp,
+    expiration: Timestamp,
+) -> Result<Record<N, Rrsig<Octs, N>>, SigningError>
+where
+    N: ToName + Clone + Send,
+    D: RecordData
+        + ComposeRecordData
+        + From<Dnskey<Octs>>
+        + CanonicalOrd
+        + Send,
+    Inner: SignRaw,
+    Octs: AsRef<[u8]> + OctetsFrom<Vec<u8>>,
+{
+    sign_rrset_in(key, rrset, apex_owner, inception, expiration, &mut vec![])
+}
+
+/// Generate `RRSIG` records for a given RRset.
+///
+/// This function generates an `RRSIG` record for the given RRset based on the
+/// given signing key, according to the rules defined in [RFC 4034 section 3]
+/// _"The RRSIG Resource Record"_, [RFC 4035 section 2.2] _"Including RRSIG
+/// RRs in a Zone"_ and [RFC 6840 section 5.11] _"Mandatory Algorithm Rules"_.
+///
+/// No checks are done on the given signing key, any key with any algorithm,
+/// apex owner and flags may be used to sign the given RRset.
+///
+/// When signing multiple RRsets by calling this function multiple times, the
+/// `scratch` buffer parameter can be allocated once and re-used for each call
+/// to avoid needing to allocate the buffer for each call.
+///
+/// [RFC 4034 section 3]:
+///     https://www.rfc-editor.org/rfc/rfc4034.html#section-3
+/// [RFC 4035 section 2.2]:
+///     https://www.rfc-editor.org/rfc/rfc4035.html#section-2.2
+/// [RFC 6840 section 5.11]:
+///     https://www.rfc-editor.org/rfc/rfc6840.html#section-5.11
+pub fn sign_rrset_in<N, D, Octs, Inner>(
+    key: &SigningKey<Octs, Inner>,
+    rrset: &Rrset<'_, N, D>,
+    apex_owner: &N,
+    inception: Timestamp,
+    expiration: Timestamp,
+    scratch: &mut Vec<u8>,
+) -> Result<Record<N, Rrsig<Octs, N>>, SigningError>
+where
+    N: ToName + Clone + Send,
+    D: RecordData
+        + ComposeRecordData
+        + From<Dnskey<Octs>>
```

Comment:
From\<Dnskey> is not needed and confusing

Reply:
Good catch. It was needed in an earlier version of the code but indeed is no longer needed.